### PR TITLE
restart_process: add custom_build_with_restart

### PR DESCRIPTION
**Pull request by nicks**
Hello @maiamcc,

Please review the following commits I made in branch nicks/restart:

e1576155b4fab12aa7c6e62b5a02969a325364c6 (2020-10-30 13:16:32 -0400)
restart_process: add custom_build_with_restart

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

File: restart_process/README.md [hunk 1]
```diff
@@ -1,6 +1,13 @@
...
+
+- `docker_build_with_restart`: wraps a `docker_build` call
+- `custom_build_with_restart`: wraps a `custom_build` call
+
+At the end of a `live_update`, the container's process will rerun itself.
+
+(Use it in place of the `restart_container()` Live Update step, which has been deprecated for Kubernetes resources.)
 
 ## When to Use
 Use this extension when you have an image (via `docker_build`) and you want to re-execute its entrypoint/command as part of a `live_update`.
```
Comment (maiamcc) [inline restart_process/README.md:13]:
```suggestion
Use this extension when you have an image and you want to re-execute its entrypoint/command as part of a `live_update`.
```

File: restart_process/README.md [hunk 2]
```diff
@@ -70,6 +90,27 @@ def docker_build_with_restart(ref: str, context: str,
...
+    """
+     Args:
+      ref: name for this image (e.g. 'myproj/backend' or 'myregistry/myproj/backend'); as the parameter of the same name in docker_build
+      command: build command for building your image
+      deps: source dependencies of the custom build
+      entrypoint: the command to be (re-)executed when the container starts or when a live_update is run
+      live_update: set of steps for updating a running container; as the parameter of the same name in docker_build
+      base_suffix: suffix for naming the base image, applied as {ref}{base_suffix}
+      restart_file: file that Tilt will update during a live_update to signal the entrypoint to rerun
+      **kwargs: will be passed to the underlying `docker_build` call
```
Comment (maiamcc) [inline restart_process/README.md:112]:
```suggestion
      **kwargs: will be passed to the underlying `custom_build` call
```

File: restart_process/Tiltfile [hunk 3]
```diff
@@ -74,5 +50,73 @@ def docker_build_with_restart(ref, context, entrypoint, live_update,
     # to timestamp changes, only writes (see https://github.com/eradman/entr/issues/32)
     live_update = live_update + [run('date > {}'.format(restart_file))]
 
+    # We don't need a real context. See:
+    # https://github.com/tilt-dev/tilt/issues/3897
+    context=ext_dir
```
Comment (maiamcc) [inline restart_process/Tiltfile:55]: good call! gah why didn't I think of this?

File: restart_process/Tiltfile [hunk 4]
```diff
@@ -74,5 +50,73 @@ def docker_build_with_restart(ref, context, entrypoint, live_update,
...
+    """Wrap a custom_build call and its associated live_update steps so that the last step
+    of any live update is to rerun the given entrypoint.
+
+
+     Args:
+      ref: name for this image (e.g. 'myproj/backend' or 'myregistry/myproj/backend'); as the parameter of the same name in docker_build
+      command: build command for building your image
+      deps: source dependencies of the custom build
+      entrypoint: the command to be (re-)executed when the container starts or when a live_update is run
+      live_update: set of steps for updating a running container; as the parameter of the same name in docker_build
```
Comment (maiamcc) [inline restart_process/Tiltfile:101]:
```suggestion
      live_update: set of steps for updating a running container; as the parameter of the same name in custom_build
```

File: restart_process/Tiltfile [hunk 5]
```diff
@@ -74,5 +50,73 @@ def docker_build_with_restart(ref, context, entrypoint, live_update,
...
+
+
+def custom_build_with_restart(ref, command, deps, entrypoint, live_update,
+                              base_suffix='-tilt_docker_build_with_restart_base', restart_file=RESTART_FILE, **kwargs):
+    """Wrap a custom_build call and its associated live_update steps so that the last step
+    of any live update is to rerun the given entrypoint.
+
+
+     Args:
+      ref: name for this image (e.g. 'myproj/backend' or 'myregistry/myproj/backend'); as the parameter of the same name in docker_build
```
Comment (maiamcc) [inline restart_process/Tiltfile:97]:
```suggestion
      ref: name for this image (e.g. 'myproj/backend' or 'myregistry/myproj/backend'); as the parameter of the same name in custom_build
```

File: restart_process/Tiltfile [hunk 6]
```diff
@@ -74,5 +50,73 @@ def docker_build_with_restart(ref, context, entrypoint, live_update,
...
+                 "`restart_container()` (this extension is meant to REPLACE restart_container() )")
+
+    for k, v in kwargs.items():
+        if k == 'skips_local_docker':
+            fail("`custom_build_with_restart` is not compatible with `skips_local_docker`, because it needs access to the image")
+
+    # rename the original image to make it a base image and declare a docker_build for it
+    base_ref = '{}{}'.format(ref, base_suffix)
+    custom_build(base_ref, command=command, deps=deps, **kwargs)
+    _helper(base_ref, ref, entrypoint, live_update, restart_file, **kwargs)
```
Comment (maiamcc) [inline restart_process/Tiltfile:122]: you need to take out any kwargs that are custom_build-specific before passing them through here, or else the `docker_build` call in `_helper` will choke

**Review by maiamcc — COMMENTED** [review]
[the `tag` param gave me trouble last time i tried this](https://github.com/tilt-dev/tilt-extensions/blob/restart_proc_custom_build/restart_process/Tiltfile#L110) -- do you have a solution? if not, `custom_build_with_restart` should disallow that param.

File: restart_process/README.md (hunk 1 above)
Comment (nicks) [reply restart_process/README.md:13]: done

File: restart_process/README.md (hunk 2 above)
Comment (nicks) [reply restart_process/README.md:112]: done

File: restart_process/Tiltfile (hunk 5 above)
Comment (nicks) [reply restart_process/Tiltfile:97]: done

File: restart_process/Tiltfile (hunk 4 above)
Comment (nicks) [reply restart_process/Tiltfile:101]: done

File: restart_process/Tiltfile (hunk 6 above)
Comment (nicks) [reply restart_process/Tiltfile:122]: done

**Review by nicks — COMMENTED** [review]
not totally sure what problem you were having with the tag param? i added a test for it and it seemed to work OK

**Comment by nicks** [comment]
oh, wait, now i remember why you can't use this with the tag= param.

Back story: DM and I decided later that adding the tag= param was a huge mistake

For all legitimate use-cases, the right way to handle this is to tack " && docker tag X $EXPECTED_REF" to the end of your custom_build script, to put it in the right place

But adding the tag param opened up a bunch of illegitimate use-cases where people were trying to hack it to do something weird, that works _sometimes_ , but breaks a lot in other cases, and so those people ended up with a broken tilt experience : \

**Review by maiamcc — APPROVED** [review]
🙌 

(btw, could you push code review changes as separate commits so I can look at just changes since my last review? ty!)

**Comment by nicks** [comment]
ya, really wish git code review handled the diffs even on an amend!

**Pull request merged**